aii-shellfe cannot verify server certificates when fetching profiles

In Quattor 17.2, aii-shellfe cannot get profiles because of a certificate verification error:
```
> aii-shellfe --configure XXXX
[INFO] No existing cache /tmp/aii/XXXX/data/aHR0cHM6Ly9xdWF0dG9yMS5waHlzaWsucnd0aC1hYWNoZW4uZGUvcHJvZmlsZXMvaG9tZTMtc2w3LnhtbA==, not specifying the modification date while retrieving
[WARN] Got an unexpected result while retrieving https://xxxxx/profiles/XXXX.xml: 500 Can't connect to xxxxx:443 (SSL connect attempt failed with unknown errorerror:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed)
```

ncm-aiiserver writes entries for `ca_file`, `cert_file`, and `key_file` into /etc/aii/aii-shellfe.conf, but the temporary configuration of ccm-fetch (/tmp/aii/XXXX/ccm.conf) does not contain these entries. Therefore the new (compared to Quattor 16.8) download method cannot verify the certificate of the server.

In my case, a workaround is to change `cdburl` in /etc/aii/aii-shellfe.conf from `https://xxxx/...` to `file:///....` in order to access the file locally.

Here are the debug ouput and some configuration files:

```
# aii-shellfe --configure xxxxx --debug 5
[DEBUG] Nodes for remove: 
[DEBUG] removelist: 
[DEBUG] Using Net::HTTPS SSL_SOCKET_CLASS IO::Socket::SSL
[DEBUG] Using LWP::UserAgent version 5.833
[DEBUG] LWP::UserAgent is recent enough to support verify_hostname for IO::Socket::SSL
[DEBUG] Using LWP::UserAgent ssl_opts SSL_ca_file: /etc/sindes/certs/ca-quattor.server.some.where.crt SSL_cert_file: /etc/sindes/certs/client_cert.pem SSL_key_file: /etc/sindes/certs/client_privatekey.pem verify_hostname: 1
[DEBUG] update_env PERL_LWP_SSL_VERIFY_HOSTNAME = 1.
[DEBUG] update_env delete PERL_NET_HTTPS_SSL_SOCKET_CLASS.
[DEBUG] update_env PERL_LWP_SSL_VERIFY_HOSTNAME = 1.
[DEBUG] update_env delete PERL_NET_HTTPS_SSL_SOCKET_CLASS.
[DEBUG] Downloading profiles-info: https://quattor.server.some.where/profiles/profiles-info.xml
[DEBUG] Parsing XML file from https://quattor.server.some.where/profiles/profiles-info.xml
[DEBUG] Added xxxxx to the list
[DEBUG] Nodes for configure: xxxxx
[DEBUG] configurelist: xxxxx
[DEBUG] Fetching profile: https://quattor.server.some.where/profiles/xxxxx.xml
[VERB] Opening file /tmp/aii/xxxxx/ccm.conf
[VERB] Running the command: /sbin/restorecon /tmp/aii/xxxxx/ccm.conf
/sbin/restorecon:  Warning no default label for /tmp/aii/xxxxx/ccm.conf
[VERB] File /tmp/aii/xxxxx/ccm.conf was modified
[DEBUG] config file /tmp/aii/xxxxx/ccm.conf  changed.
[DEBUG] base_url is not defined in configuration
[DEBUG] URL is https://quattor.server.some.where/profiles/xxxxx.xml
[DEBUG] No lockfile /tmp/aii/xxxxx/fetch.lock found: no lock
[DEBUG] flock on /tmp/aii/xxxxx/fetch.lock gave has_lock 1
[DEBUG] Writing global lock /tmp/aii/xxxxx/global.lock
[VERB] Opening file /tmp/aii/xxxxx/global.lock
[VERB] Running the command: /sbin/restorecon /tmp/aii/xxxxx/global.lock
/sbin/restorecon:  Warning no default label for /tmp/aii/xxxxx/global.lock
[VERB] File /tmp/aii/xxxxx/global.lock was modified
[VERB] Opening file /tmp/aii/xxxxx/latest.cid
[DEBUG] No reference file/pipe via source option. Returning false.
[VERB] Opening file /tmp/aii/xxxxx/profile.0/profile.url
[DEBUG] No reference file/pipe via source option. Returning false.
[VERB] Not saving file /tmp/aii/xxxxx/profile.0/profile.url
[VERB] Opening file /tmp/aii/xxxxx/profile.0/profile.xml
[DEBUG] No reference file/pipe via source option. Returning false.
[VERB] Not saving file /tmp/aii/xxxxx/profile.0/profile.xml
[DEBUG] Current URL https://quattor.server.some.where/profiles/xxxxx.xml is different from the previous fetched one . Forcing download.
[INFO] No existing cache /tmp/aii/xxxxx/data/aHR0cHM6Ly9xdWF0dG9yMS5waHlzaWsucnd0aC1hYWNoZW4uZGUvcHJvZmlsZXMvYW1zLXduMDUueG1s, not specifying the modification date while retrieving
[VERB] FORCE set, not setting if_modified_since in request
[DEBUG] Using Net::HTTPS SSL_SOCKET_CLASS IO::Socket::SSL
[DEBUG] Using LWP::UserAgent version 5.833
[DEBUG] LWP::UserAgent is recent enough to support verify_hostname for IO::Socket::SSL
[DEBUG] Using LWP::UserAgent ssl_opts verify_hostname: 1
[DEBUG] update_env PERL_LWP_SSL_VERIFY_HOSTNAME = 1.
[DEBUG] update_env delete PERL_NET_HTTPS_SSL_SOCKET_CLASS.
[DEBUG] update_env PERL_LWP_SSL_VERIFY_HOSTNAME = 1.
[DEBUG] update_env delete PERL_NET_HTTPS_SSL_SOCKET_CLASS.
[WARN] Got an unexpected result while retrieving https://quattor.server.some.where/profiles/xxxxx.xml: 500 Can't connect to quattor.server.some.where:443 (SSL connect attempt failed with unknown errorerror:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed)
[DEBUG] https://quattor.server.some.where/profiles/xxxxx.xml: try 1 of 3: sleeping for 30 seconds
[VERB] FORCE set, not setting if_modified_since in request
[DEBUG] Using Net::HTTPS SSL_SOCKET_CLASS IO::Socket::SSL
[DEBUG] Using LWP::UserAgent version 5.833
[DEBUG] LWP::UserAgent is recent enough to support verify_hostname for IO::Socket::SSL
[DEBUG] Using LWP::UserAgent ssl_opts verify_hostname: 1
[DEBUG] update_env PERL_LWP_SSL_VERIFY_HOSTNAME = 1.
[DEBUG] update_env delete PERL_NET_HTTPS_SSL_SOCKET_CLASS.
[DEBUG] update_env PERL_LWP_SSL_VERIFY_HOSTNAME = 1.
[DEBUG] update_env delete PERL_NET_HTTPS_SSL_SOCKET_CLASS.
[WARN] Got an unexpected result while retrieving https://quattor.server.some.where/profiles/xxxxx.xml: 500 Can't connect to quattor.server.some.where:443 (SSL connect attempt failed with unknown errorerror:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed)
[DEBUG] https://quattor.server.some.where/profiles/xxxxx.xml: try 2 of 3: sleeping for 30 seconds
[VERB] FORCE set, not setting if_modified_since in request
[DEBUG] Using Net::HTTPS SSL_SOCKET_CLASS IO::Socket::SSL
[DEBUG] Using LWP::UserAgent version 5.833
[DEBUG] LWP::UserAgent is recent enough to support verify_hostname for IO::Socket::SSL
[DEBUG] Using LWP::UserAgent ssl_opts verify_hostname: 1
[DEBUG] update_env PERL_LWP_SSL_VERIFY_HOSTNAME = 1.
[DEBUG] update_env delete PERL_NET_HTTPS_SSL_SOCKET_CLASS.
[DEBUG] update_env PERL_LWP_SSL_VERIFY_HOSTNAME = 1.
[DEBUG] update_env delete PERL_NET_HTTPS_SSL_SOCKET_CLASS.
[WARN] Got an unexpected result while retrieving https://quattor.server.some.where/profiles/xxxxx.xml: 500 Can't connect to quattor.server.some.where:443 (SSL connect attempt failed with unknown errorerror:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed)
[DEBUG] https://quattor.server.some.where/profiles/xxxxx.xml: try 3 of 3: sleeping for 30 seconds
[ERROR] Failed to fetch profile https://quattor.server.some.where/profiles/xxxxx.xml
[VERB] Running the command: /sbin/restorecon /tmp/aii/xxxxx/latest.cid
/sbin/restorecon:  Warning no default label for /tmp/aii/xxxxx/latest.cid
[VERB] File /tmp/aii/xxxxx/latest.cid was modified
[ERROR] Impossible to fetch profile for xxxxx: failed to download. Skipping.
[ERROR] No nodes left to process after checking for protected hosts
[DEBUG] closing down
```

ccm.conf used by aii-shellfe:
```
# cat /tmp/aii/xxxxx/ccm.conf 
cache_root /tmp/aii/xxxxx
json_typed 1
tabcompletion 0
```

For comparision: ccm.conf generated by ncm-ccm:
```
# cat /etc/ccm.conf
ca_file /etc/sindes/certs/ca-quattor.server.some.where.crt
cache_root /var/lib/ccm
cert_file /etc/sindes/certs/client_cert.pem
debug 0
force 0
get_timeout 30
key_file /etc/sindes/certs/client_privatekey.pem
lock_retries 3
lock_wait 30
profile https://quattor.server.some.where/profiles/xxxxx.xml
retrieve_retries 3
retrieve_wait 30
world_readable 0
```

Configuration of aii-shellfe:
```
# cat /etc/aii/aii-shellfe.conf 
# File generated by ncm-aiiserver
# Do not edit
ca_file = /etc/sindes/certs/ca-quattor.server.some.where.crt
#cdburl = file:///opt/profiles
cdburl = https://quattor.server.some.where/profiles
cert_file = /etc/sindes/certs/client_cert.pem
key_file = /etc/sindes/certs/client_privatekey.pem
nbpdir = /data/aii/nbp/pxelinux.cfg
osinstalldir = /data/aii/www/ks
profile_format = xml
use_fqdn = 0
```

--- a/aii-core/src/main/perl/Shellfe.pm
+++ b/aii-core/src/main/perl/Shellfe.pm
@@ -378,6 +378,40 @@ sub _initialize
     return $self;
 }
 
+
+# Extract CAF::Download::LWP or CCM HTTPS (or other)
+# download related options from aii-shellfe config/options
+# Returns a hashref with options.
+# type can be lwp or ccm
+sub _download_options
+{
+    my ($self, $type) = @_;
+
+    my @related_opts = (CERT, KEY, CAFILE, CADIR);
+
+    # map AII config option names to the LWP ones
+    my $map = {
+        lwp => {
+            CERT() => 'cert',
+            KEY() => 'key',
+            CAFILE() => 'cacert',
+            CADIR() => 'cadir',
+        },
+    };
+
+    my $opts = {}; # make sure it can always be used as %$opts
+    foreach my $optname (@related_opts) {
+        my $val = $self->option($optname);
+        if ($val) {
+            my $key = $map->{$type}->{$optname} || $optname;
+            $opts->{$key} = $val;
+        }
+    };
+
+    return $opts;
+}
+
+
 # Lock a node being configured, needs to be called in every method that contains
 # node operations (ie configure etc)
 sub lock_node
@@ -584,15 +618,8 @@ sub nodelist
             my $url = $self->option (CDBURL) . "/" . PROFILEINFO;
 
             my $lwp = CAF::Download::LWP->new(log => $self);
-            my %lwp_opts_map = (
-                CERT() => 'cert',
-                KEY() => 'key',
-                CAFILE() => 'cacert',
-                CADIR() => 'cadir'
-                );
-            my %lwp_opts = map {$lwp_opts_map{$_} => $self->option($_)} grep {$self->option($_)} keys %lwp_opts_map;
-
-            my $rp = $lwp->_do_ua('get', [$url], %lwp_opts);
+            my $opts = $self->_download_options('lwp');
+            my $rp = $lwp->_do_ua('get', [$url], %$opts);
             $self->debug (4, "Downloading profiles-info: $url");
             unless ($rp->is_success) {
                 $self->error ("Couldn't download $url. Aborting ",
@@ -700,6 +727,11 @@ sub fetch_profiles
             print $cfg_fh "json_typed $json_typed\n";
             print $cfg_fh "tabcompletion 0\n";
 
+            my $opts = $self->_download_options('ccm');
+            foreach my $optname (sort keys %$opts) {
+                print $cfg_fh "$optname $opts->{$optname}\n";
+            }
+
             my $changed = $cfg_fh->close() ? "" : "not";
             $self->debug(1, "config file $config $changed changed.");
         };
@@ -744,7 +776,7 @@ sub fetch_profiles
 }
 
 # Initiate the Parallel:ForkManager with requested threads if option is given
-sub init_pm 
+sub init_pm
 {
     my ($self, $cmd, $responses) = @_;
     if ($self->option('parallel')) {
@@ -755,10 +787,10 @@ sub init_pm
                 if ($exit_code) {
                     $self->error("Error running $cmd on $id, exitcode $exit_code");
                 };
-                # retrieve data structure from child 
+                # retrieve data structure from child
                 if (defined($data_struct_ref)) {
                     $responses->{$id} = $data_struct_ref;
-                    $self->debug(5, "Running $cmd on $id had output"); 
+                    $self->debug(5, "Running $cmd on $id had output");
                 }
             }
         );
@@ -772,7 +804,7 @@ sub init_pm
 sub run_cmd {
     my ($self, $cmd, %node_states) = @_;
     my $method = "_$cmd";
-    my %responses; 
+    my %responses;
     my $pm = $self->init_pm($cmd, \%responses);
     foreach my $node (sort keys %node_states) {
         $self->debug (2, "$cmd: $node");
@@ -786,7 +818,7 @@ sub run_cmd {
         $self->debug(5, "Going to start $cmd on node $node");
         if ($pm) {
             if ($pm->start($node)){
-                $self->verbose("started execution for $node in child"); 
+                $self->verbose("started execution for $node in child");
                 next;
             } else {
                 $self->verbose("child with parallel execution for $node");
@@ -794,7 +826,7 @@ sub run_cmd {
         }
         my $ec = $self->$method($node, $node_states{$node}) || 0;
         my $res = { ec => $ec, method => $method, node => $node, mode => $pm ? 1 : 0 };
-    
+
         if ($pm) {
             $self->verbose("Terminating the child for $node");
             $pm->finish($ec, $res);
@@ -947,26 +979,26 @@ sub check_protected {
     return %hash;
 }
 
-sub change_dhcp 
+sub change_dhcp
 {
     my ($self, $method, %nodes) = @_;
     $self->debug(5, "logfile:", $self->option('logfile'), " dhcpcfg:", $self->option('dhcpcfg'));
     my $dhcpmgr = AII::DHCP->new('script',
-        "--logfile=".$self->option('logfile'), 
-        "--cfgfile=".$self->option('dhcpcfg'), 
+        "--logfile=".$self->option('logfile'),
+        "--cfgfile=".$self->option('dhcpcfg'),
         log => $self);
     foreach my $node (sort keys %nodes) {
-        my $st = $nodes{$node}; 
+        my $st = $nodes{$node};
         $self->debug(3, "Checking dhcp config on node $node for method $method.");
         if ($st->{configuration}->elementExists(DHCPPATH)) {
             $self->dhcp($node, $st, $method, $dhcpmgr);
         }
-    }    
+    }
     $dhcpmgr->configure_dhcp();
 
     return 1;
 }
-      
+
 
 # Runs all the commands
 sub cmds
@@ -1048,7 +1080,7 @@ sub cmds
                 $self->verbose('got option NODHCP, DHCP not enabled');
             } else {
                 # If needed, do DHCP here
-                if ("$cmd" eq "configure"){ 
+                if ("$cmd" eq "configure"){
                     $self->change_dhcp(CONFIGURE, %nodes);
                 } elsif ("$cmd" eq "remove"){
                     $self->change_dhcp(REMOVEMETHOD, %nodes);

--- a/aii-core/src/test/perl/shellfe.t
+++ b/aii-core/src/test/perl/shellfe.t
@@ -1,0 +1,49 @@
+use strict;
+use warnings;
+use Test::More;
+use Test::Quattor;
+use AII::Shellfe;
+
+use Readonly;
+use File::Basename qw(basename);
+
+my $logfile_name = basename(__FILE__);
+$logfile_name =~ s/\.t$//;
+
+# log dir
+mkdir 'target/test';
+
+
+Readonly my $AII_CONFIG_EMPTY => 'src/test/resources/aii-empty.conf';
+Readonly my $AII_CONFIG_CERTS => 'src/test/resources/aii-certs.conf';
+Readonly my $AII_LOG_FILE => "target/test/$logfile_name.log";
+Readonly::Array my @SHELLFE_DEFAULT_OPTIONS => ('--logfile', $AII_LOG_FILE);
+
+my $cli = AII::Shellfe->new('name',
+                            @SHELLFE_DEFAULT_OPTIONS,
+                            '--cfgfile', $AII_CONFIG_CERTS,
+                            '--cert', '/some/other/path/to/cert');
+
+is_deeply($cli->_download_options('lwp'), {
+    cacert => 'test/ca_file',
+    cadir => 'test/ca_dir',
+    key => 'test/key_file',
+    cert => '/some/other/path/to/cert',
+}, "_download_options converted aii options in CAF::Download::LWP options");
+
+is_deeply($cli->_download_options('ccm'), {
+    ca_file => 'test/ca_file',
+    ca_dir => 'test/ca_dir',
+    key_file => 'test/key_file',
+    cert_file => '/some/other/path/to/cert',
+}, "_download_options converted aii options in CCM options");
+
+# Test empty config
+# must return empty hashref (so %$res works)
+$cli = AII::Shellfe->new('name',
+                         @SHELLFE_DEFAULT_OPTIONS,
+                         '--cfgfile', $AII_CONFIG_EMPTY);
+is_deeply($cli->_download_options('lwp'), {}, "empty config returns hashref for lwp");
+is_deeply($cli->_download_options('ccm'), {}, "empty config returns hashref for ccm");
+
+done_testing;

--- a/aii-core/src/test/resources/aii-certs.conf
+++ b/aii-core/src/test/resources/aii-certs.conf
@@ -1,0 +1,4 @@
+ca_file = test/ca_file
+ca_dir = test/ca_dir
+key_file = test/key_file
+cert_file = test/cert_file

